Product release 1/update pricing

anitaa1990 opened the pull request:
Fixes #1774. This PR adds the option to update the product pricing info.

**Note that this feature is only enabled in debug builds, so end users won't see it yet.**

~- Adding the `Not ready for merge` label since this [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1495) needs to be merged before this PR can be merged.~

#### Screenshots
<img width="300" src="https://user-images.githubusercontent.com/22608780/74737093-2183e080-527a-11ea-8695-5a2c9a01adf8.gif">

#### Testing
- Update the product pricing details and click on Update and verify that the Product pricing is updated. - Schedule a sale or cancel a scheduled sale and verify that it is updated correctly.
- Verify that discard is working as expected.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

anitaa1990 commented:
Adding the `Not ready for merge` label since this [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1495) needs to be merged before this PR can be merged.

nbradbury commented:
If I change a value on the pricing screen then change it back to what it was, I still see the discard dialog even though nothing has changed.

![Screenshot_1582047728](https://user-images.githubusercontent.com/3903757/74762727-6a24a500-524c-11ea-81e6-d5944c936feb.png)

nbradbury commented:
I tried to schedule sale from Feb 18 - 20, but when I select Feb 20 as the end date and tap OK, the end date shows up as a day earlier.

I suspect GMT conversion is involved here. Note that the device & store are in the same time zone (GMT - 5:00).

![1](https://user-images.githubusercontent.com/3903757/74763119-31d19680-524d-11ea-8cac-2bbc7b6f4c17.png)

![2](https://user-images.githubusercontent.com/3903757/74763127-35651d80-524d-11ea-9ccf-460a5163dbc2.png)

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt:
```diff
@@ -166,7 +187,14 @@ class ProductDetailViewModel @AssistedInject constructor(
+        regularPrice: BigDecimal? = null,
+        salePrice: BigDecimal? = null,
+        isSaleScheduled: Boolean? = null,
+        dateOnSaleFrom: String? = null,
```
Are `dateOnSaleFrom` and `dateOnSaleTo` not GMT?

anitaa1990 replied:
They are! Fixed in d84deae

nbradbury commented:
If I background the app while the pricing fragment is showing then restore it, I get this crash:

```
ava.lang.IllegalArgumentException: Required value was null.
        at com.woocommerce.android.ui.products.ProductPricingFragment.updateProductTaxClassList(ProductPricingFragment.kt:212)
        at com.woocommerce.android.ui.products.ProductPricingFragment.access$updateProductTaxClassList(ProductPricingFragment.kt:34)
        at com.woocommerce.android.ui.products.ProductPricingFragment$setupObservers$1.invoke(ProductPricingFragment.kt:108)
        at com.woocommerce.android.ui.products.ProductPricingFragment$setupObservers$1.invoke(ProductPricingFragment.kt:34)
```

nbradbury commented:
@anitaa1990 Done with my first round!

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailRepository.kt:
```diff
@@ -121,6 +148,9 @@ class ProductDetailRepository @Inject constructor(
             productStore.getVariationsForProduct(selectedSite.get(), remoteProductId).size
 
+    fun getTaxClassesForSite(): List<TaxClass> =
+            taxStore.getShippingClassListForSite(selectedSite.get()).map { it.toAppModel() }
```
Just noticed this FluxC routine is named `getShippingClassListForSite`. Shouldn't it be `getTaxClassesForSite()`?

anitaa1990 replied:
Good catch! I have opened a [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1499) to update that. Once that is merged, will update the hash in this PR 👍

anitaa1990 replied:
Updated FluxC hash in f738742

anitaa1990 commented:
Thanks for the review @nbradbury!

> If I change a value on the pricing screen then change it back to what it was, I still see the discard dialog even though nothing has changed.

Good catch! Fixed in  98bd13c.

> I tried to schedule sale from Feb 18 - 20, but when I select Feb 20 as the end date and tap OK, the end date shows up as a day earlier.

I was able to reproduce this by changing my device timezone to GMT - 12. Added a fix in 23ffec3.

> If I background the app while the pricing fragment is showing then restore it, I get this crash:

I noticed that this crash happens in both `ProductInventoryFragment` and `ProductPricingFragment` when `Do not keep activities` is enabled. I have opened a separate issue in #1967 to tackle this separately since the fix would most likely involve making changes to both those screens and this PR is big enough as is. WDYT?

Unrelated, I found a different crash happening when the app was resumed from the background when in the Product Detail screen and fixed it in 6d9c5c1.

PR Ready for another round!

nbradbury commented:
> I noticed that this crash happens in both `ProductInventoryFragment` and `ProductPricingFragment` when `Do not keep activities` is enabled. I have opened a separate issue in #1967 to tackle this separately since the fix would most likely involve making changes to both those screens and this PR is big enough as is. WDYT?

I'm fine with that. Fwiw, I think the problem is simply due to using `val product = requireNotNull(productData.product)` when the product is null .

nbradbury commented:
The bug where the discard dialog was being shown after changing a field to it's previous value appears to be fixed for everything but the "schedule sale" toggle. If you enable that then disable it again, you still see the discard dialog. I think that can be handled separately, though, since as you've mentioned this PR  is already quite large.

It doesn't look like the GMT sale date problem is resolved, though. As these screenshots illustrate, I set the sale dates from Feb 20 - 22 and tapped done, and product detail showed the dates as Feb 19 - 22. Tapping to see the pricing screen again also showed Feb 19 - 22.

![1](https://user-images.githubusercontent.com/3903757/74930635-14670e80-53ac-11ea-9f8a-fef6c823e35c.png)

nbradbury commented:
> Unrelated, I found a different crash happening when the app was resumed from the background when in the Product Detail screen and fixed it in [6d9c5c1](https://github.com/woocommerce/woocommerce-android/commit/6d9c5c1d00301c74a895018c49d80488593d5c6a).

Thanks for that! I noticed the same parcelable crash yesterday but didn't have time to look into it yet.

anitaa1990 commented:
> I'm fine with that. Fwiw, I think the problem is simply due to using val product = requireNotNull(productData.product) when the product is null .

Yup, that's the reason. But it looks like when the activity is recreated, both screens are calling the `product` model before it can be fetched which doesn't seem very ideal and and I thought it would be nice to explore other ways to achieve that.

> The bug where the discard dialog was being shown after changing a field to it's previous value appears to be fixed for everything but the "schedule sale" toggle. If you enable that then disable it again, you still see the discard dialog. 

Good catch! This was because when sale is switched on for the first time, then the start and the end date is set to  the current date and when it was switched off again, the dates were not reseted back. Fixed in d0bcd51

> It doesn't look like the GMT sale date problem is resolved, though. As these screenshots illustrate, I set the sale dates from Feb 20 - 22 and tapped done, and product detail showed the dates as Feb 19 - 22. Tapping to see the pricing screen again also showed Feb 19 - 22.

Ahh, my bad! I didn't include the time when calculating the sale dates so by default it was being set to `00:00:00`.Fixed in 96584f9

Ready for another round 👍

nbradbury commented:
> Good catch! This was because when sale is switched on for the first time, then the start and the end date is set to the current date and when it was switched off again, the dates were not reseted back. Fixed in [d0bcd51](https://github.com/woocommerce/woocommerce-android/commit/d0bcd5198aeb9da52da40d0baaa47eacaec458b1)

This doesn't seem to be fixed. If I have schedule sale toggled off and then toggle it on and off again, I still get the Discard confirmation dialog. But I think we're fine to tackle that separately -if it even needs to be tackled, since it's a nit pick rather than a real problem.
